server/src/main/java/org/opensearch/querygroup/LabelingService.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.querygroup;
+
+import org.opensearch.action.IndicesRequest;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.plugins.LabelingPlugin;
+
+import java.util.EnumMap;
+import java.util.List;
+
+/**
+ * TODO: Don't know the right package to put this class, need suggestions for maintainers on this
+ * Main class to hold multiple implementations of {@link org.opensearch.plugins.LabelingPlugin}
+ * this class will facilitate access and interactions to different implementations
+ * Usage: This class should be used as a member to orchestrate the working of your {@link LabelingPlugin}
+ */
+public class LabelingService {
+    /**
+     * Enum to define what all are currently implementing the plugin
+     */
+    public enum LabelingImplementationType {
+        QUERY_GROUP_RESOURCE_MANAGEMENT,
+        NOOP
+    }
+
+    EnumMap<LabelingImplementationType, LabelingPlugin> implementations;
+
+    public LabelingService(List<LabelingPlugin> loadedPlugins) {
+        implementations = new EnumMap<>(LabelingImplementationType.class);
+        for (LabelingPlugin plugin : loadedPlugins) {
+            if (implementations.containsKey(plugin.getImplementationName())) {
+                throw new IllegalArgumentException("There should not be two implementations of a LabelingImplementation type");
+            }
+            implementations.put(plugin.getImplementationName(), plugin);
+        }
+    }
+
+    /**
+     * populates the threadContext with the labels yielded by the {@param type} against the {@link LabelingHeader} keys
+     * @param type
+     * @param request
+     * @param threadContext
+     */
+    public void labelRequestFor(final LabelingImplementationType type, final IndicesRequest request, final ThreadContext threadContext) {
+        final LabelingPlugin plugin = implementations.get(type);
+        if (plugin == null) {
+            throw new IllegalArgumentException(type + " implementation is not enabled");
+        }
+        plugin.labelRequest(request, threadContext);
```

Comment:
I'm still in favor of having multiple plugins labeling a requests at the same time like https://github.com/opensearch-project/OpenSearch/pull/14282/files#diff-f02ca59c8d99a70fdc277b1b311ba8fc0135920ef5a114578187e9388e922f6bR60, 
Can we define something like a "order/priority" that a plugin can apply the labels like we did in the `ActionFilter`: https://github.com/opensearch-project/OpenSearch/blob/112704bf2976db8b0dd71980173896029a490cd9/server/src/main/java/org/opensearch/action/support/ActionFilter.java#L50

Reply:
What you are thinking of is I think from a different perspective, I think you are assuming that all of the consuming feature needs are uniform but they are not as I had explained it above for QueryGrouping feature.

So rather than **push-down-to-syn** mechanism to apply this event. I am more inclined towards **pull-to-sync** mechanism

Reply:
I see, could you also give some examples on where this function will be called? if we are using this on demand, we might need to pass `labelingService` to all plugins that need it? possibly by adding it as a default constructor parameter for `LabelingPlugin` ?